File: packages/analysis/src/stemmer-sv.ts
import {
  longestSuffixInRegion,
  markScandinavianRegion,
} from "./stemmer-scandinavian.js";

const VOWELS = new Set(["a", "e", "i", "o", "u", "y", "å", "ä", "ö"]);
const S_ENDING = new Set([
  "b",
  "c",
  "d",
  "f",
  "g",
  "h",
  "j",
  "k",
  "l",
  "m",
  "n",
  "o",
  "p",
  "r",
  "t",
  "v",
  "y",
]);
const OST_ENDING = new Set(["i", "k", "l", "n", "p", "r", "t", "u", "v"]);
const ET_EXCEPTIONS = [
  "h",
  "iet",
  "uit",
  "fab",
  "cit",
  "dit",
  "alit",
  "ilit",
  "mit",
  "nit",
  "pit",
  "rit",
  "sit",
  "tit",
  "ivit",
  "kvit",
  "xit",
  "kom",
  "rak",
  "pak",
  "stak",
];
const MAIN_SUFFIXES = [
  "a",
  "arna",
  "erna",
  "heterna",
  "orna",
  "ad",
  "e",
  "ade",
  "ande",
  "arne",
  "are",
  "aste",
  "en",
  "anden",
  "aren",
  "heten",
  "ern",
  "ar",
  "er",
  "heter",
  "or",
  "as",
  "arnas",
  "ernas",
  "ornas",
  "es",
  "ades",
  "andes",
  "ens",
  "arens",
  "hetens",
  "erns",
  "at",
  "andet",
  "het",
  "ast",
  "s",
  "et",
];

function validEtPrefix(prefix: string): boolean {
  if (prefix.length < 3) return false;
  const last = prefix[prefix.length - 1] as string;
  const beforeLast = prefix[prefix.length - 2] as string;
  if (VOWELS.has(last) || !VOWELS.has(beforeLast)) return false;
  return !ET_EXCEPTIONS.some((suffix) => prefix.endsWith(suffix));
}

function mainSuffix(word: string, p1: number): string {
  const suffix = longestSuffixInRegion(word, p1, MAIN_SUFFIXES);
  if (suffix === undefined) return word;
  if (suffix === "s") {
    const beforeS = word.slice(0, -1);
    if (beforeS.endsWith("et") && validEtPrefix(beforeS.slice(0, -2))) {
      return beforeS.slice(0, -2);
    }
    return S_ENDING.has(beforeS.at(-1) ?? "") ? beforeS : word;
  }
  if (suffix === "et") {
    const prefix = word.slice(0, -2);
    return validEtPrefix(prefix) ? prefix : word;
  }
  return word.slice(0, -suffix.length);
}

function consonantPair(word: string, p1: number): string {
  const suffix = longestSuffixInRegion(word, p1, [
    "dd",
    "gd",
    "nn",
    "dt",
    "gt",
    "kt",
    "tt",
  ]);
  return suffix === undefined ? word : word.slice(0, -1);
}

function otherSuffix(word: string, p1: number): string {
  const suffix = longestSuffixInRegion(word, p1, [
    "lig",
    "ig",
    "els",
    "öst",
    "fullt",
  ]);
  if (suffix === undefined) return word;
  if (suffix === "fullt") return `${word.slice(0, -suffix.length)}full`;
  if (suffix === "öst") {
    const prefix = word.slice(0, -suffix.length);
    return OST_ENDING.has(prefix.at(-1) ?? "") ? `${prefix}ös` : word;
  }
  return word.slice(0, -suffix.length);
}

/** Snowball Swedish stemmer, verified against snowball-data. */
export function stemSwedish(word: string): string {
  if (word.length === 0) return word;
  const p1 = markScandinavianRegion(word, VOWELS);
  return otherSuffix(consonantPair(mainSuffix(word, p1), p1), p1);
}

File: python/searchable-analysis/src/searchable_analysis/stemmer_sv.py
from searchable_analysis.stemmer_scandinavian import (
    longest_suffix_in_region,
    mark_scandinavian_region,
)

VOWELS = frozenset("aeiouyåäö")
S_ENDING = frozenset("bcdfghjklmnoprtvy")
OST_ENDING = frozenset("iklnprtuv")
ET_EXCEPTIONS = (
    "h",
    "iet",
    "uit",
    "fab",
    "cit",
    "dit",
    "alit",
    "ilit",
    "mit",
    "nit",
    "pit",
    "rit",
    "sit",
    "tit",
    "ivit",
    "kvit",
    "xit",
    "kom",
    "rak",
    "pak",
    "stak",
)
MAIN_SUFFIXES = (
    "a",
    "arna",
    "erna",
    "heterna",
    "orna",
    "ad",
    "e",
    "ade",
    "ande",
    "arne",
    "are",
    "aste",
    "en",
    "anden",
    "aren",
    "heten",
    "ern",
    "ar",
    "er",
    "heter",
    "or",
    "as",
    "arnas",
    "ernas",
    "ornas",
    "es",
    "ades",
    "andes",
    "ens",
    "arens",
    "hetens",
    "erns",
    "at",
    "andet",
    "het",
    "ast",
    "s",
    "et",
)


def _valid_et_prefix(prefix: str) -> bool:
    if len(prefix) < 3 or prefix[-1] in VOWELS or prefix[-2] not in VOWELS:
        return False
    return not prefix.endswith(ET_EXCEPTIONS)


def _main_suffix(word: str, p1: int) -> str:
    suffix = longest_suffix_in_region(word, p1, MAIN_SUFFIXES)
    if suffix is None:
        return word
    if suffix == "s":
        before = word[:-1]
        if before.endswith("et") and _valid_et_prefix(before[:-2]):
            return before[:-2]
        return before if before[-1:] in S_ENDING else word
    if suffix == "et":
        return word[:-2] if _valid_et_prefix(word[:-2]) else word
    return word[: -len(suffix)]


def _consonant_pair(word: str, p1: int) -> str:
    suffix = longest_suffix_in_region(word, p1, ("dd", "gd", "nn", "dt", "gt", "kt", "tt"))
    return word if suffix is None else word[:-1]


def _other_suffix(word: str, p1: int) -> str:
    suffix = longest_suffix_in_region(word, p1, ("lig", "ig", "els", "öst", "fullt"))
    if suffix is None:
        return word
    prefix = word[: -len(suffix)]
    if suffix == "fullt":
        return prefix + "full"
    if suffix == "öst":
        return prefix + "ös" if prefix[-1:] in OST_ENDING else word
    return prefix


def stem_swedish(word: str) -> str:
    if not word:
        return word
    p1 = mark_scandinavian_region(word, VOWELS)
    return _other_suffix(_consonant_pair(_main_suffix(word, p1), p1), p1)
